File: answers/21/main.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
import sys
from answers.hist import *

def hist_normalize(img, a, b):
    c, d = img.min(), img.max()
    # if c <= xin < d
    out = (b - a) / (d - c) * (img - c) + a
    # if xin < c
    out[img < c] = a
    # if xin > d
    out[img > d] = b
    return np.clip(out, 0, 255).astype(np.uint8)

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256_dark.png')
    img_result = hist_normalize(img_orig, a=0, b=256)

    hist_orig = makeHisto(img_orig, 256, 0, 256)
    hist_result = makeHisto(img_result, 256, 0, 256)

    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.imshow('python_origin_hist', hist_orig)
    cv2.imshow('python_result_hist', hist_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.moveWindow('python_origin_hist', img_orig.shape[1]+img_result.shape[1], 0)
    cv2.moveWindow('python_result_hist', img_orig.shape[1]+img_result.shape[1]+hist_orig.shape[1], 0)
    cv2.waitKey(0)

File: answers/21/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>
#include "../hist.cpp"

void hist_normalize(const cv::Mat &src, cv::Mat &dst, float a, float b) {
    dst = src.clone();
    dst.convertTo(dst, CV_32F);

    double c, d;
    cv::minMaxLoc(src, &c, &d);
    if (c == d) {
        return;
    }

    // if c <= xin < d
    dst -= cv::Scalar::all(c);
    dst *= 255 / (d - c);
    dst.convertTo(dst, CV_8U);

    // if xin < c
    dst *= (b - a) / 255;
    dst += cv::Scalar::all(a);
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256_dark.png");
    cv::Mat img_result;
    hist_normalize(img_orig, img_result, 0, 256);

    cv::Mat hist_orig, hist_result;
    makeHisto(img_orig, hist_orig, 256, 0, 256);
    makeHisto(img_result, hist_result, 256, 0, 256);
    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::imshow("cpp_origin_hist", hist_orig);
    cv::imshow("cpp_result_hist", hist_result);
    cv::moveWindow("cpp_origin", 0, -img_orig.size().height);
    cv::moveWindow("cpp_result", img_orig.size().width, -img_orig.size().height);
    cv::moveWindow("cpp_origin_hist", img_orig.size().width + img_result.size().width, -img_orig.size().height);
    cv::moveWindow("cpp_result_hist", img_orig.size().width + img_result.size().width + hist_orig.size().width, -img_orig.size().height);
    cv::waitKey(0);
    return 0;
}
